### PR TITLE
Library Migration | `rainbow kit` and `wagmi`

### LINKED ISSUE #895: Wallet Provider Update / Migration
Research doc:

https://www.notion.so/hulking-cosmos-3e8/Wallet-Provider-Update-617df4fce464459b89bbd801820ed56c

### PR DESCRIPTION
**Pull request by Da-Colon**
## Description

Wallet Provider Swap out. `web3modal` v0.1.8 that we were previously using has been depreciated. Upgrading to use `rainbowkit` and `wagmi` libraries to take care of the connection to wallets

Supports:
- InjectedWallets
- WalletConnect

## Notes
- I will follow up with another PR to remove left over code and library
- I've replaced all instances of `Web3Provider` with customs hooks from `wagmi`
- There are some future opportunities to remove our implementation of address ens lookup with `wagmi` hooks down the road

## Issue / Notion doc (if applicable)
Closes #895

### CODE REVIEW

File: src/components/ConnectWalletToast.tsx [hunk 1]
```diff
@@ -25,9 +26,7 @@ function ToastContent({ label }: { label: string }) {
 }
 
 function ConnectWalletToast({ label }: { label: string }) {
-  const {
-    state: { account },
-  } = useWeb3Provider();
+  const { address: account } = useAccount();
```
Comment (Da-Colon) [inline src/components/ConnectWalletToast.tsx:29]: You will see I use `{address: account}` and `{data: signer}` to make it a little clearer what the variables are when used through out the files

File: src/components/Proposals/MultisigProposalDetails/TxActions.tsx [hunk 2]
```diff
@@ -26,9 +27,10 @@ export function TxActions({
     gnosis: { safe, safeService },
     actions: { refreshSafeData },
   } = useFractal();
-  const {
-    state: { account, signerOrProvider, chainId },
-  } = useWeb3Provider();
+  const provider = useProvider();
+  const { data: signer } = useSigner();
+  const signerOrProvider = useMemo(() => signer || provider, [signer, provider]);
```
Comment (Da-Colon) [inline src/components/Proposals/MultisigProposalDetails/TxActions.tsx:32]: This one pattern is repeated quite a bit. but as it currently set up to read/write we need this to update when a user is connected as a signer. Currently `Wagmi` does not combine these into a single variable

File: src/components/Proposals/MultisigProposalDetails/TxActions.tsx [hunk 3]
```diff
@@ -50,7 +52,7 @@ export function TxActions({
       asyncRequest({
         asyncFunc: () =>
           (signerOrProvider as Signer & TypedDataSigner)._signTypedData(
-            { verifyingContract: safe.address, chainId },
+            { verifyingContract: safe.address, chainId: provider._network.chainId },
```
Comment (Da-Colon) [inline src/components/Proposals/MultisigProposalDetails/TxActions.tsx:55]: So the `wagmi` hook `useNetwork` is normally what you would use to read the currently connected `chain` information. Currently this wallet is set to `undefined` when a wallet is not connected. While the app will remain connected to a fallback provider. 

To get around this to allow for not only typescript but for the places to where we need this information when a wallet is not connect `provider._network.chainId` is used.

File: src/constants/url.ts [hunk 4]
```diff
@@ -1,3 +1,3 @@
 export const URL_FAQ = 'https://docs.fractalframework.xyz/welcome-to-fractal/user-guides/faq';
 export const URL_DISCORD = 'https://discord.com/invite/decent-dao';
-export const URL_DOCS = 'https://docs.fractalframework.xyz/welcome-to-fractal';
+export const URL_DOCS = 'https://docs.fractalframework.xyz/welcome-to-fractal/';
```
Comment (Da-Colon) [inline src/constants/url.ts:3]: grr

File: src/providers/NetworkConfig/NetworkConfigProvider.tsx [hunk 5]
```diff
@@ -41,14 +43,27 @@ const getNetworkConfig = (chainId: number) => {
...
+    const supportedChainIds =
+      process.env.REACT_APP_SUPPORTED_CHAIN_IDS?.split(',').map(id => parseInt(id)) || [];
+
+    if (!!chain && !supportedChainIds.includes(chain.id)) {
+      toast(t('toastSwitchChain', { chainNames: supportedChainIds }), {
+        toastId: 'switchChain',
+      });
+      disconnect();
+    }
+  }, [chain, disconnect, t]);
```
Comment (Da-Colon) [inline src/providers/NetworkConfig/NetworkConfigProvider.tsx:66]: This handles non supported networks by disconnecting the wallet. if a user clicks connect wallet on a unsupported network, they are prompted to chain to supported network.

File: src/providers/NetworkConfig/rainbow-kit.config.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  infuraProvider({ priority: 0, apiKey: process.env.REACT_APP_INFURA_API_KEY! }),
+  alchemyProvider({ priority: 1, apiKey: process.env.REACT_APP_ALCHEMY_API_KEY! }),
+  publicProvider({ priority: 2 }),
+]);
+
+const defaultWallets = [injectedWallet({ chains }), walletConnectWallet({ chains })];
+// allows connection to localhost only in development mode.
+if (process.env.REACT_APP_TESTING_ENVIROMENT) {
+  defaultWallets.unshift(testWallet({ chains }));
+}
```
Comment (Da-Colon) [inline src/providers/NetworkConfig/rainbow-kit.config.ts:28]: This testWallet connects during tests automatically via this conditional.

File: src/providers/NetworkConfig/testWallet.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+      chains,
+      options: {
+        signer: new providers.JsonRpcProvider().getSigner(),
+      },
+    });
+    return {
+      connector,
+    };
+  },
+});
```
Comment (Da-Colon) [inline src/providers/NetworkConfig/testWallet.ts:25]: This is never shown a user, for testing environment only

File: tests/models/NavPage.ts [hunk 8]
```diff
@@ -6,79 +6,60 @@ export enum MenuItems {
...
     }
   }
 
-  async clickMenuConnect() {
-    await this.page.click('[data-testid=accountMenu-connect]');
-  }
-
   async clickMenuDisconnect() {
-    await this.page.click('[data-testid=accountMenu-disconnect]');
+    await this.page.click('[data-testid="accountMenu-disconnect"]');
```
Comment (Da-Colon) [inline tests/models/NavPage.ts:27]: Noticed these were missing quotes at least according to docs. I had some test failing and was exhausting all means so I updated. This probably won't matter comes @herbig tests update

**Review by Da-Colon — COMMENTED** [review]
Don't let the size full you. should be a pretty quick review. Let me know if you have any questions

File: src/providers/Fractal/FractalProvider.tsx [hunk 9]
```diff
@@ -120,5 +121,5 @@ export function FractalProvider({ children }: { children: ReactNode }) {
     ]
   );
 
-  return <FractalContext.Provider value={value}>{children}</FractalContext.Provider>;
+  return <FractalContext.Provider value={value as any}>{children}</FractalContext.Provider>;
```
[the remark this answers is not in the document]
Comment (Da-Colon) [reply src/providers/Fractal/FractalProvider.tsx:124]: ah part of debugging the swap. I'll get that removed.

File: src/constants/url.ts (hunk 4 above)
Comment (herbig) [reply src/constants/url.ts:3]: I hate it

File: src/components/Proposals/MultisigProposalDetails/TxActions.tsx (hunk 2 above)
Comment (herbig) [reply src/components/Proposals/MultisigProposalDetails/TxActions.tsx:33]: can you put these three lines into a separate hook itself, and call that everywhere and get the data you need?

**Review by herbig — APPROVED** [review]
🚢

**Comment by tbwebb22** [comment]
image

Is "injected wallet" what we want to show to a user? Just wondering if most users will know what that means.

**Pull request merged**